GEECS-PythonAPI/geecs_python_api/controls/data_acquisition/scan_manager.py

```diff
@@ -336,6 +365,152 @@ def fill_async_nans(self, log_df, async_observables, fill_value=0):
+            except Exception as e:
+                logging.error(f"Error extracting timestamp from {device_file}: {e}")
+
+        return matched_rows
+
+    def extract_timestamp_from_file(self, device_file, device_type):
+        """
+        Extract timestamp from a device file based on its type.
+        """
+        device_map = {
+            "Point Grey Camera": get_imaq_timestamp_from_png,
+            "MagSpecCamera": get_imaq_timestamp_from_png,
+            "PicoscopeV2": get_picoscopeV2_timestamp,
+            "MagSpecStitcher": get_magspecstitcher_timestamp
+        }
+
+        if device_type in device_map:
+            return device_map[device_type](device_file)
+        else:
+            raise ValueError(f"Unsupported device type: {device_type}")
+
+    def process_dependent_directories(self, device_name, device_dir, matched_rows, scan_number):
+        """
+        Process and rename files in dependent directories.
+        """
+        for suffix in self.DEPENDENT_SUFFIXES:
+            dependent_dir = device_dir.parent / f"{device_name}{suffix}"
+            if dependent_dir.exists() and dependent_dir.is_dir():
+                logging.info(f"Processing dependent directory: {dependent_dir}")
+                dependent_files = list(dependent_dir.glob("*"))
+                self.rename_files_in_dependent_directory(dependent_files, matched_rows, scan_number, suffix)
+
+    def rename_files_in_dependent_directory(self, dependent_files, matched_rows, scan_number, suffix):
+        """
+        Rename dependent files based on matched rows from the master directory.
+        """
+        for i, (master_file, row_index) in enumerate(matched_rows):
+            if i < len(dependent_files):
+                dependent_file = dependent_files[i]
+                master_new_name = re.sub(r'_\d+$', '', master_file.stem)  # Remove trailing number
+                new_name = f"{scan_number}_{master_new_name}{suffix}_{str(row_index + 1).zfill(3)}{dependent_file.suffix}"
+                new_path = dependent_file.parent / new_name
+                dependent_file.rename(new_path)
+                logging.info(f"Renamed {dependent_file} to {new_path}")
+            else:
+                logging.warning(f"Not enough files in dependent directory to match {master_file}")
+
+    def rename_files(self, matched_rows, scan_number, device_name):
+        """
+        Rename master files based on scan number, device name, and matched row index.
+        """
+        for file_path, row_index in matched_rows:
+            row_number = str(row_index + 1).zfill(3)
+            new_file_name = f"{scan_number}_{device_name}_{row_number}{file_path.suffix}"
+            new_file_path = file_path.parent / new_file_name
+            if not new_file_path.exists():
+                logging.info(f"Renaming {file_path} to {new_file_path}")
+                os.rename(file_path, new_file_path)
```

Comment:
Was this `os` here because it is more compatible with the `ThreadPoolExecutor()` of line 402 than `pathlib.Path` for multithreaded file operations?  Otherwise, could consider changing to a `Path` to entirely remove `os`

Reply:
yeah, I tend to agree. Originally, I had some other code here to automatically determine relevant paths based off specific machine details (domain, computer name etc.). I switched to this config file based approach so as to not have specific machine details in the code. 

The load_config function in geecs_database was added by me some time ago when we first introduced this config file. But it's not really the right spot for it. Probably it makes sense to consolidate/move the stuff in geecs_database and here to geecs_python_api/tools/interface. But, I'll leave that as a future task.

Reply:
switched to pathlib